packages/apollo-language-server/src/diagnostics.ts

```diff
@@ -11,25 +13,43 @@ import { GraphQLDocument } from "./document";
 import { highlightNodeForNode } from "./utilities/graphql";
 import { rangeForASTNode } from "./utilities/source";
 
+import { TypeKind } from "apollo/lib/printer/ast";
 import { getValidationErrors } from "apollo/lib/validation";
 
-export function collectDiagnostics(
+/**
+ * Build an array of code diagnostics for all GraphQL operations in a document.
+ * This function explicitly ignores type definitions.
+ */
+export function collectOperationDiagnostics(
   schema: GraphQLSchema,
   queryDocument: GraphQLDocument,
   fragments: { [fragmentName: string]: FragmentDefinitionNode }
 ): Diagnostic[] {
   const ast = queryDocument.ast;
   if (!ast) return queryDocument.syntaxErrors;
 
+  // I would prefer to use the TypeKind enum from graphql-js here,
+  // but the enum values aren't the same, i.e. "OBJECT" vs "ObjectTypeDefinition"
+  const typeKinds = Object.values(TypeKind);
+  const astWithoutTypeDefinitions = visit(ast, {
+    enter(node: ASTNode) {
```

Comment:
Note that `visit` also allows you to specify specific nodes to visit, so you don't have to perform a manual check. See [here](https://github.com/apollographql/apollo-cli/blob/c50691d8c8fdb0e2beea817052c28ccf5887ea8a/packages/apollo-language-server/src/project.ts#L469) for example. Or [here](https://github.com/graphql/graphql-js/blob/26c9874107e65f19576aae0a32638287820f68aa/src/language/visitor.js#L142) for the `visit` docs, which describe all usage patterns.

Reply:
In this case however, a visitor might be overkill. You can simply loop over `ast.definitions` and either filter out type definitions and extensions, or only keep operation and fragment definitions.

Reply:
I think this could actually be simplified to `ast.definitions.map(isExecutableDefinitionNode)`!

Reply:
I like that a lot! Thanks for the info and the links, I'll take a look.